src/core/server/saved_objects/migrations/core/transform_saved_object_document_error.ts

```diff
@@ -0,0 +1,49 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0 and the Server Side Public License, v 1; you may not use this file except
+ * in compliance with, at your election, the Elastic License 2.0 or the Server
+ * Side Public License, v 1.
+ */
+
+/**
+ * Error thrown when saved object migrations encounter a transformation error.
+ * Transformation errors happen when a transform function throws an error for an unsanitized saved object
+ * The id (doc.id) reported in this error class is just the uuid part and doesn't tell users what the full elasticsearch id is.
+ * in order to convert the id to the serialized version further upstream using serializer.generateRawId, we need to provide the following items:
+ * - namespace: doc.namespace,
+ * - type: doc.type,
+ * - id: doc.id,
+ * The new error class helps with v2 migrations.
+ * For backward compatibility with v1 migrations, the error message is the same as what was previouslt thrown as a plain error
+ */
+
+// TINA TODO: create getters for retrieving the id, type and namespace to use in migrate_raw_doc for generating the serialized SO id
+export class TransformSavedObjectDocumentError extends Error {
+  constructor(
+    public readonly id: string,
+    public readonly type: string,
+    public readonly namespace: string,
+    public readonly failedTransform: string, // created by document_migrator wrapWithTry as `${type.name}:${version}`;
+    public readonly failedDoc: string,
+    public readonly originalError: Error // public readonly namespace?: string--> part of the stringified failedDoc
+  ) {
+    super(`Failed to transform document ${id}. Transform: ${failedTransform}\nDoc: ${failedDoc}`);
+    this.id = id;
+    this.namespace = namespace;
+    this.type = type;
+    // Removed because not including still seems to work, it may have been an old Typescript 2.1 issue:
```

Comment:
Ideally, we also want to capture the stack trace. @pgayvallet do you happen to know if I need to add something along the lines of:
```suggestion
    // Maintains proper stack trace for where our error was thrown (only available on V8)
    if (Error.captureStackTrace) {
      Error.captureStackTrace(this, TransformSavedObjectDocumentError)
    }
    // Removed because not including still seems to work, it may have been an old Typescript 2.1 issue:
```
as suggested in the "ES6 Custom Error Class" https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Error#instance_properties section?

I was kinda assuming we'd get the stack trace along with the error.

Reply:
` Error.captureStackTrace` captures the stacktrace at the time it was invoked, so when the `TransformSavedObjectDocumentError` will be created. Is that what you want?

If you want to use the stack from the original error instead (`originalError`), You can just copy it in the constructor (`this.stack = originalError.stack`), or even access it from consumer code, as the property is public `transformError.originalError.stack`

Reply:
I want the stack trace from the original error because that's ultimately what's adding to the reason the migration will fail.

Reply:
I tried to add the stack traces to the `FATAL` controlState reason but realized it's going to clutter up the logs badly! For now, we'll go with adding the raw id and the transform that failed. We can tack on the stack trace later if we see that it's not enough info for debugging failed migrations.

Reply:
maybe we can improve it in a follow-up